api/smp_utils

```diff
@@ -20,38 +20,85 @@
 #ifndef API_SMP_UTILS_HEADER
 #define API_SMP_UTILS_HEADER
 
-#include <arch.hpp>
+#include <emmintrin.h>
+#include <chrono>
 
-/// x86-related locking stuff ///
-#if defined(ARCH_x86)
-// Intel 3a  8.10.6.7: 128-byte boundary
-typedef unsigned int spinlock_t __attribute__((aligned(128)));
+/* Architecture independent TTAS[1] spinlock implementation.
+ *
+ * Implements C++ named requirements BasicLockable, Lockable and TimedLockable
+ * for use with e.g. std::lock_guard.
+ *
+ * Loosely follows C++ spinlock implementation by Erik Rigtorp[2].
+ *
+ * References:
+ * 1. https://en.wikipedia.org/wiki/Test_and_test-and-set
+ * 2. https://rigtorp.se/spinlock/
+ * */
+class Spinlock {
+  public:
+    void lock() noexcept {
+      // See https://rigtorp.se/spinlock/ for details on memory ordering
+      //
+      // Optimised for uncontended locks as it starts with exchange, which
+      // requires a write.
+      while (lock_.exchange(true, std::memory_order_acquire)) {
+        while (lock_.load(std::memory_order_relaxed)) {
+          _mm_pause();
```

Comment:
`_mm_pause()` is purely x86 intrinsic, and won't work on arm where you have `__yield()` instead. It's basically a hint to the internal CPU scheduler, to aid hyperthreading.

Reply:
Ah, yes - then I guess we still have to include an architecture check here.

Reply:
I'm thinking about ways to implement this check. To try to avoid platform-specific checks outside the architecture specific include files, perhaps we could introduce an `os::Arch::cpu_relax` for each platform? Here: https://github.com/includeos/IncludeOS/blob/master/api/arch/x86_64.hpp#L19

We could then implement it as `_mm_pause` for x86 and skip the aarch64 implementation for now (so it would give a build error later) or just add it as an untested call to `__yield`.

What do you think @alfreb @fwsGonzo ?

Reply:
That makes sense to me if the semantics of yield and pause are similar enough that they can use the same argument.

Reply:
On the other hand - we don't actively support arm yet, and this lock should work just fine on arm as well without the intrinsic, so for me it's fine to do this later as well.